src/Shrubbery/Union.hs

```diff
@@ -91,6 +98,56 @@ dissectUnion ::
 dissectUnion branches (Union branchIndex t) =
   selectBranchAtIndex branchIndex branches t
 
+{- |
+  Matches a 'Union' against one of its types, returning 'Just' the value if the
+  'Union' value is of that type, or 'Nothing' otherwise. Use with @TypeApplications@:
+
+  @
+    matchUnion @Int someUnion
+  @
+
+  Note that this will always match the first instance of the type in the union types list.
+
+  See 'matchUnionProxy' for a version that does not require @TypeApplications@.
+-}
+matchUnion ::
+  forall t types branchIndex.
+  ( KnownNat branchIndex
+  , branchIndex ~ FirstIndexOf t types
+  ) =>
+  Union types ->
+  Maybe t
+matchUnion (Union branchIndex t) =
+  let
+    indexOfT :: BranchIndex t types
+    indexOfT = firstIndexOfType @branchIndex
+  in
+    case testBranchIndexEquality indexOfT branchIndex of
+      Just Refl -> Just t
+      Nothing -> Nothing
+
+{- |
+  Matches a 'Union' against one of its types, returning 'Just' the value if the
+  'Union' value is of that type, or 'Nothing' otherwise, e.g.:
+
+  @
+    matchUnionProxy (Proxy :: ProxyInt) myUnion
+  @
+
+  Note that this will always match the first instance of the type in the union types list.
+
+  See 'matchUnion' for a version that uses @TypeApplications@ instead of a proxy.
+-}
+matchUnionProxy ::
+  forall t types branchIndex proxy.
+  ( KnownNat branchIndex
+  , branchIndex ~ FirstIndexOf t types
+  ) =>
+  proxy t ->
+  Union types ->
+  Maybe t
+matchUnionProxy _ = matchUnion
```

Comment:
Does using `Proxy t`  instead of `proxy t` work here?

Reply:
Yes but there's no reason to use `Proxy t` instead.

Reply:
It'd be one less type parameter which is nice though right? Or do you lose some generality somehow

Reply:
You lose the ability to use a carrier type other than `Proxy`. It's not a big deal in practice, but the rest of the library follows this pattern. See for example [`selectBranchAtProxy`](https://github.com/flipstone/shrubbery/blob/b84c9f1b869136e76aa9300762bd02fd2de0cadd/src/Shrubbery/Branches.hs#L137).

Reply:
Ah okay cool, I see